packages/firestore/src/core/sync_engine.ts

```diff
@@ -958,9 +964,42 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
           targetId,
           /*current=*/ false
         );
+      } else {
+        // For queries that have a local View, we need to update their state
+        // in LocalStore (as the resume token and the snapshot version
+        // might have changed) and reconcile their views with the persisted
+        // state (the list of syncedDocuments may have gotten out of sync).
+        await this.localStore.releaseTarget(
+          targetId,
+          /*keepPersistedQueryData=*/ true
+        );
+
+        assert(
+          !!this.queriesByTarget[targetId] &&
+            !this.queriesByTarget[targetId].isEmpty(),
+          `There is no queries mapped to target id ${targetId}`
+        );
+
+        const queries = this.queriesByTarget[targetId].toArray();
+        queryData = await this.localStore.allocateTarget(queries[0].toTarget());
+
+        for (const query of queries) {
+          const queryView = this.queryViewsByQuery.get(query);
+          assert(!!queryView, `No query view found for ${query}`);
+
+          const viewChange = await this.synchronizeViewAndComputeSnapshot(
+            queryView!
+          );
+          if (viewChange.snapshot) {
+            newViewSnapshots.push(viewChange.snapshot);
+          }
+        }
       }
-      activeQueries.push(queryData);
+
+      assert(!!queryData, `Target data for target ${targetId} not found`);
```

Comment:
I slightly prefer doing the wrong thing here and changing the queryData type back to just `QueryData` (rather than `QueryData|undefined`. queryData is always assigned to above and you could get rid of the assert, which is also slightly misleading since the LocalStore is meant to allocate QueryData, rather than retrieve it.